src/Components/Components/src/ComponentFactory.cs

```diff
@@ -49,23 +49,16 @@ private static ComponentTypeInfoCacheEntry GetComponentTypeInfo([DynamicallyAcce
     public IComponent InstantiateComponent(IServiceProvider serviceProvider, [DynamicallyAccessedMembers(Component)] Type componentType, IComponentRenderMode? callerSpecifiedRenderMode, int? parentComponentId)
     {
         var (componentTypeRenderMode, propertyInjector) = GetComponentTypeInfo(componentType);
-        IComponent component;
 
-        if (componentTypeRenderMode is null && callerSpecifiedRenderMode is null)
-        {
-            // Typical case where no rendermode is specified in either location. We don't call ResolveComponentForRenderMode in this case.
-            component = _componentActivator.CreateInstance(componentType);
-        }
-        else
-        {
-            // At least one rendermode is specified. We require that it's exactly one, and use ResolveComponentForRenderMode with it.
-            var effectiveRenderMode = callerSpecifiedRenderMode is null
-                ? componentTypeRenderMode!
-                : componentTypeRenderMode is null
-                    ? callerSpecifiedRenderMode
-                    : throw new InvalidOperationException($"The component type '{componentType}' has a fixed rendermode of '{componentTypeRenderMode}', so it is not valid to specify any rendermode when using this component.");
-            component = _renderer.ResolveComponentForRenderMode(componentType, parentComponentId, _componentActivator, effectiveRenderMode);
-        }
+        // In the typical case where no rendermode is specified in either location, we don't even call
+        // ComputeEffectiveRenderMode and hence don't call ResolveComponentForRenderMode either.
+        var effectiveRenderMode = componentTypeRenderMode is null && callerSpecifiedRenderMode is null
+            ? null
+            : _renderer.ResolveEffectiveRenderMode(componentType, parentComponentId, componentTypeRenderMode, callerSpecifiedRenderMode);
+
+        var component = effectiveRenderMode is null
+            ? _componentActivator.CreateInstance(componentType)
+            : _renderer.ResolveComponentForRenderMode(componentType, parentComponentId, _componentActivator, effectiveRenderMode);
```

Comment:
Conceptually the change here is factoring out the "resolve rendermode" logic into a new `virtual` method on the renderer, so that we can implement the new rule about static pages suppressing callsite rendermodes. Doing so happens to simplify `ComponentFactory` quite a bit.